vault/plugin_catalog.go

```diff
@@ -34,20 +39,135 @@ type PluginCatalog struct {
 	lock sync.RWMutex
 }
 
-func (c *Core) setupPluginCatalog() error {
+func (c *Core) setupPluginCatalog(ctx context.Context) error {
 	c.pluginCatalog = &PluginCatalog{
 		builtinRegistry: c.builtinRegistry,
 		catalogView:     NewBarrierView(c.barrier, pluginCatalogPath),
 		directory:       c.pluginDirectory,
 	}
 
+	err := c.pluginCatalog.UpgradePlugins(ctx, c.logger)
+	if err != nil {
+		c.logger.Error("error while upgrading plugin storage", "error", err)
+	}
+
 	if c.logger.IsInfo() {
 		c.logger.Info("successfully setup plugin catalog", "plugin-directory", c.pluginDirectory)
 	}
 
 	return nil
 }
 
+func (c *PluginCatalog) UpgradePlugins(ctx context.Context, logger log.Logger) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// If the directory isn't set we can skip the upgrade attempt
+	if c.directory == "" {
+		return nil
+	}
+
+	// List plugins from old location
+	pluginsRaw, err := c.catalogView.List(ctx, "")
+	if err != nil {
+		return err
+	}
+	plugins := make([]string, 0, len(pluginsRaw))
+	for _, p := range pluginsRaw {
+		if !strings.HasSuffix(p, "/") {
+			plugins = append(plugins, p)
+		}
+	}
+
+	logger.Info("upgrading plugin information", "plugins", plugins)
+
+	var retErr error
+	for _, pluginName := range plugins {
+		pluginRaw, err := c.catalogView.Get(ctx, pluginName)
+		if err != nil {
+			return err
+		}
+
+		plugin := new(pluginutil.PluginRunner)
+		if err := jsonutil.DecodeJSON(pluginRaw.Value, plugin); err != nil {
+			return errwrap.Wrapf("failed to decode plugin entry: {{err}}", err)
```

Comment:
Should this `continue` instead, so one bad decode doesn't stop all subsequent plugins?

Reply:
updated